File: Documentacao/LIA2_008_PedagogicalAdaptive_Mobile_Thinking_FIX02/apps/backend/app/contracts/academicStageContract.py
from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class AcademicStageCreateContract(BaseModel):
    contractName: Literal["AcademicStageCreate.v1"] = "AcademicStageCreate.v1"
    educationLevel: str = Field(min_length=2, max_length=80)
    stageCode: str | None = Field(default=None, max_length=80)
    stageLabel: str = Field(min_length=2, max_length=160)
    startedAt: date | None = None
    endedAt: date | None = None
    status: Literal["CURRENT", "COMPLETED", "CANCELLED"] = "CURRENT"

    @model_validator(mode="after")
    def validateDates(self):
        if (
            self.startedAt is not None
            and self.endedAt is not None
            and self.endedAt < self.startedAt
        ):
            raise ValueError("endedAt não pode ser anterior a startedAt.")

        return self


class AcademicStageContract(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    contractName: Literal["AcademicStage.v1"] = "AcademicStage.v1"
    academicStageId: UUID
    studentId: UUID
    educationLevel: str
    stageCode: str | None
    stageLabel: str
    startedAt: date | None
    endedAt: date | None
    status: Literal["CURRENT", "COMPLETED", "CANCELLED"]
    createdAt: datetime
    updatedAt: datetime

File: apps/studentWeb/src/contracts/academicStageContract.ts
export type AcademicStageStatus = 'CURRENT' | 'COMPLETED' | 'CANCELLED'

export interface AcademicStageContract {
  contractName: 'AcademicStage.v1'
  academicStageId: string
  studentId: string
  educationLevel: string
  stageCode: string | null
  stageLabel: string
  startedAt: string | null
  endedAt: string | null
  status: AcademicStageStatus
  createdAt: string
  updatedAt: string
}

export interface AcademicStageCreateContract {
  contractName: 'AcademicStageCreate.v1'
  educationLevel: string
  stageCode: string | null
  stageLabel: string
  startedAt: string | null
  endedAt: string | null
  status: AcademicStageStatus
}
